### ISSUE TITLE
How to consume Decorator Components in ES6

### ISSUE THREAD

**Issue by veenedu** [palantir/blueprint#925]
Hi,

I am using blueprintjs with  [facebook's create-react-app](https://github.com/facebookincubator/create-react-app) . I want to consume blueprint's HotkeysTarget. As per [docs](http://blueprintjs.com/docs/#components.hotkeys.api) we have to add decorator to our Component class.

As of now create-react-app doesn't support decorators, [read me](https://github.com/facebookincubator/create-react-app/blob/master/packages/react-scripts/template/README.md#can-i-use-decorators)

So how can I use HotkeysTarget and similar other component

**Comment by giladgray** [comment]
so decorators are just functions with special syntax that invokes them for you, but they're very easy to unwind and call yourself. your "read me" link above links to two issues that provide acceptable workarounds for decorators in environments that don't support them.

[this](https://github.com/facebookincubator/create-react-app/issues/411#issuecomment-238974268) is the easiest way.

**Comment by giladgray** [comment]
should be along the lines of `export const MyHotkeysComponent = HotkeysTarget(MyComponent)` 🎉

**Comment by veenedu** [comment]
I tried both ways.

```
//Method 1
const TestHotKeysComponent1 =  HotkeysTarget(TestComponent);

//TestHotKeysComponent1 is undefined 
console.log(TestHotKeysComponent1);


//Method 2
const TestHotKeysComponent2 =  HotkeysTarget(TestComponent);
//This throws an error. @HotkeysTarget-decorated class must implement `renderHotkeys`
```


Below is my `TestComponent`



```
import React, { Component } from 'react';
import {HotkeysTarget,Hotkey, Hotkeys,} from '@blueprintjs/core'

class TestComponent extends Component {
   
   //method copy pasted from blueprintJS Docs
    renderHotkeys(){
          return <Hotkeys>
            <Hotkey
                global={true}
                combo="shift + a"
                label="Be awesome all the time"
                onKeyDown={() => console.log("Awesome!")}
            />
            <Hotkey
                group="Fancy shortcuts"
                combo="shift + f"
                label="Be fancy only when focused"
                onKeyDown={() => console.log("So Fancy!")}
            />
        </Hotkeys>    
    }
    render() {        
        return (
            <div>
                Test Comp...
            </div>
        );
    }
}

const TestHotKeysComponent =  HotkeysTarget(TestComponent);

console.log(TestHotKeysComponent);

export default TestComponent;
```

**Comment by veenedu** [comment]
@giladgray 

[In this file hotkeysTarget.tsx](https://github.com/palantir/blueprint/blob/7596b0a54df1a9b5b4b54ceb91f0f8cac0b20d6f/packages/core/src/components/hotkeys/hotkeysTarget.tsx#L101)

If we add `return this` just before `tslint:enable`, it works fine.

Then we can use this component as `const TestHotKeysComponent =  HotkeysTarget(TestComponent);`


I tried modifying the dist file directly in node modules and it worked.  So I think it should work when we modify .tsx and re-run the build.

I modified this file.
`node_modules/@blueprintjs/core/dist/components/hotkeys/hotkeysTarget.js`

**Comment by giladgray** [comment]
it looks like the decorator impl actually modifies the component prototype. this is a side effect, but it's always worked fine as a decorator. 

```js
class Component ... {}

HotkeysTarget(Component);
// done. Component is now augmented to support `renderHotkeys`.
```

feel free to submit a PR with the return statement if you want that usage.

**Comment by veenedu** [comment]
@giladgray 
Hi,
I made the change and submitted. 
I dont know how to proceed further. Could you check.

**Comment by veenedu** [comment]
@giladgray 
Hi, sorry to bug you again
There is no need to return `this` as I mentioned earlier


we can consume decorators in bluprintJS as below.

```
//define a component
class OrgComponent {}

//Just pass your component to decorator. 
HotkeysTarget(OrgComponent);

//Return original content (its now mutated by decorator)
export default OrgComponent
```

Ideally HOF should not mutate the original passed argument, but in this case they are.

**Comment by adidahiya** [comment]
> Ideally HOF should not mutate the original passed argument, but in this case they are.

huh, yeah, this is a bit strange and I didn't realize our hotkeys components did this. It's not how [higher order components](https://github.com/vasanthk/react-bits/blob/master/patterns/15.higher-order-component.md) typically work, so we should consider refactoring HotkeysTarget and deprecating the existing API.

**Comment by veenedu** [comment]
@adidahiya 
[ContextMenuTarget](https://github.com/palantir/blueprint/blob/master/packages/core/src/components/context-menu/contextMenuTarget.tsx) has same problem

**Comment by DrewDennison** [comment]
Also ran into this issue and worked around it but would love to see the HOCs not mutate the component but return a new one for composing with Redux connect etc

**Comment by giladgray** [comment]
closing as we've resolved how to use invoke decorator functions directly and filed #931 to track mutation issue.

**Comment by jennykortina** [comment]
@giladgray when I try to export like this:

`export default connect(null, dispatcher)(ContextMenuTarget(Component));`

i get this error:

`Uncaught Error: You must pass a component to the function returned by connect. Instead received undefined at invariant`

I'm importing ContextMenuTarget like this:

`import { ContextMenuTarget, Menu, MenuItem } from '@blueprintjs/core';`

**Comment by adidahiya** [comment]
@jennykortina because `ContextMenuTarget` doesn't return anything. For now you can work around this by doing this:

```ts
ContextMenuTarget(Component);
// at this point, Component's prototype has been modified by the decorator
export default connect(null, dispatcher)(Component);
```

The proper long-term fix is #931.

**Comment by jennykortina** [comment]
@adidahiya thank you!

**Comment by jordymeow** [comment]
Hello guys! I have the same issue as @veenedu (@HotkeysTarget-decorated class should implement renderHotkeys), and I am using exactly the same code as he mentioned. It also seems it has been fixed but as of version 3.6.1, I still have the issue. Is there something I should know? 

Since googling about this issue brings to this page, it would be nice to have a clear and defined conclusion :) Thank you.
